### PR TITLE
Allow CodeFixes to edit closed AdditionalDocuments under global undo …

### LINKED ISSUE #3967: NRE when applying fix that modifies a non-cs/vb file, while that file is not open in the editor.
We have an analyzer that detects when a certain setting is inconsistent between two files (one C# file `Foo.cs`, and one XML file `FooManifest.xml`). Our fix provider provides a fix in `Foo.cs`, which updates the `FooManifest.xml` to match `Foo.cs`.

We found that if `FooManifest.xml` is not open in the editor, an NRE is thrown when applying the code fix, causing the analyzer to crash. The NRE is caused  when a null `Document` is passed to `GeneratedCodeRecognition.GeneratedCodeRecognitionServiceFactory.GeneratedCodeRecognitionService.IsGeneratedCode(Document document)`. Full stack trace below.

 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.GeneratedCodeRecognition.GeneratedCodeRecognitionServiceFactory.GeneratedCodeRecognitionService.IsGeneratedCode(Microsoft.CodeAnalysis.Document document) Line 25	C#
	Microsoft.VisualStudio.LanguageServices.Implementation.dll!Microsoft.VisualStudio.LanguageServices.RoslynVisualStudioWorkspace.OpenInvisibleEditor(Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.IVisualStudioHostDocument hostDocument) Line 130	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.DocumentProvider.StandardTextDocument.UpdateText(Microsoft.CodeAnalysis.Text.SourceText newText) Line 214	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.ApplyAdditionalDocumentTextChanged(Microsoft.CodeAnalysis.DocumentId documentId, Microsoft.CodeAnalysis.Text.SourceText newText) Line 771	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.ApplyProjectChanges(Microsoft.CodeAnalysis.ProjectChanges projectChanges) Line 1096	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.TryApplyChanges(Microsoft.CodeAnalysis.Solution newSolution) Line 865	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.TryApplyChanges(Microsoft.CodeAnalysis.Solution newSolution) Line 154	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.CodeActions.ApplyChangesOperation.Apply(Microsoft.CodeAnalysis.Workspace workspace, System.Threading.CancellationToken cancellationToken) Line 47	C#
 	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.CodeActions.CodeActionEditHandlerService.Apply(Microsoft.CodeAnalysis.Workspace workspace, Microsoft.CodeAnalysis.Document fromDocument, System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.CodeActions.CodeActionOperation> operations, string title, System.Threading.CancellationToken cancellationToken) Line 163	C#
 	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.Invoke.AnonymousMethod__0() Line 114	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformAction(Microsoft.CodeAnalysis.Extensions.IExtensionManager extensionManager, object extension, System.Action action) Line 22	C#
 	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.Invoke(System.Threading.CancellationToken cancellationToken) Line 89	C#
 	Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Language.Intellisense.Implementation.LightBulbSession.InvokeAction.AnonymousMethod__0(Microsoft.VisualStudio.Language.Intellisense.Utilities.IWaitContext c) Line 243	C#
 	Microsoft.VisualStudio.Editor.Implementation.dll!Microsoft.VisualStudio.Editor.Implementation.VisualStudioWaitIndicator.Wait(string title, string message, bool allowCancel, System.Action<Microsoft.VisualStudio.Language.Intellisense.Utilities.IWaitContext> action) Line 34	C#
 	Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Language.Intellisense.Implementation.LightBulbSession.InvokeAction(Microsoft.VisualStudio.Language.Intellisense.ISuggestedAction action, uint menuIndex) Line 238	C#
 	Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Language.Intellisense.Implementation.LightBulb.GetMenuItemForAction.AnonymousMethod__0() Line 356	C#
 	Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Language.Intellisense.Implementation.LightBulbCommand.Execute(object parameter) Line 35	C#
 	PresentationFramework.dll!MS.Internal.Commands.CommandHelpers.CriticalExecuteCommandSource(System.Windows.Input.ICommandSource commandSource, bool userInitiated) Line 197	C#
 	PresentationFramework.dll!System.Windows.Controls.MenuItem.InvokeClickAfterRender(object arg) Line 1419	C#
 	WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate callback, object args, int numArgs) Line 104	C#
 	WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.TryCatchWhen(object source, System.Delegate callback, object args, int numArgs, System.Delegate catchHandler) Line 37	C#
 	WindowsBase.dll!System.Windows.Threading.DispatcherOperation.InvokeImpl() Line 583	C#
 	WindowsBase.dll!System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(object state) Line 528	C#
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) Line 954	C#
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) Line 902	C#
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) Line 891	C#
 	WindowsBase.dll!System.Windows.Threading.DispatcherOperation.Invoke() Line 441	C#
 	WindowsBase.dll!System.Windows.Threading.Dispatcher.ProcessQueue() Line 2232	C#
 	WindowsBase.dll!System.Windows.Threading.Dispatcher.WndProcHook(System.IntPtr hwnd, int msg, System.IntPtr wParam, System.IntPtr lParam, ref bool handled) Line 2485	C#
 	WindowsBase.dll!MS.Win32.HwndWrapper.WndProc(System.IntPtr hwnd, int msg, System.IntPtr wParam, System.IntPtr lParam, ref bool handled) Line 341	C#
 	WindowsBase.dll!MS.Win32.HwndSubclass.DispatcherCallbackOperation(object o) Line 494	C#
 	WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate callback, object args, int numArgs) Line 104	C#
 	WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.TryCatchWhen(object source, System.Delegate callback, object args, int numArgs, System.Delegate catchHandler) Line 37	C#
 	WindowsBase.dll!System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority priority, System.TimeSpan timeout, System.Delegate method, object args, int numArgs) Line 1445	C#
 	WindowsBase.dll!MS.Win32.HwndSubclass.SubclassWndProc(System.IntPtr hwnd, int msg, System.IntPtr wParam, System.IntPtr lParam) Line 392	C#

### PR DESCRIPTION
**Pull request by dpoeschl**
…transactions

Fixes #3967

Update RoslynVisualStudioWorkspace.OpenInvisibleEditor to remove the assumption that Solution.GetDocument(DocumentId) will always work for a given IVisualStudioHostDocument's ID. GetDocument returns null on all AdditionalDocuments, causing the eventual reported NullReferenceException.

### CODE REVIEW

**Comment by dpoeschl** [comment]
Reviewers: @shyamnamboodiripad @heejaechang @mavasani @Pilchie @DustinCampbell @davkean @jasonmalinowski @balajikris @brettfo @CyrusNajmabadi @jmarolf @rchande

@shyamnamboodiripad @heejaechang @mavasani -- Is there a recommended way of writing tests for this?

**Comment by Pilchie** [comment]
Can this be tested?  Otherwise :+1:

File: src/VisualStudio/Core/Impl/RoslynVisualStudioWorkspace.cs [hunk 1]
```diff
@@ -124,8 +124,22 @@ internal override IInvisibleEditor OpenInvisibleEditor(IVisualStudioHostDocument
...
             // We need to ensure the file is saved, only if a global undo transaction is open
             var globalUndoService = this.Services.GetService<IGlobalUndoService>();
-            bool needsSave = globalUndoService.IsGlobalTransactionOpen(this);
-            bool needsUndoDisabled = needsSave && this.Services.GetService<IGeneratedCodeRecognitionService>().IsGeneratedCode(this.CurrentSolution.GetDocument(hostDocument.Id));
+            var needsSave = globalUndoService.IsGlobalTransactionOpen(this);
+
+            var needsUndoDisabled = false;
+            if (needsSave)
+            {
+                if (this.CurrentSolution.ContainsDocument(hostDocument.Id))
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Impl/RoslynVisualStudioWorkspace.cs]: Can you add a comment explaining when this can fail?

Comment (dpoeschl) [inline src/VisualStudio/Core/Impl/RoslynVisualStudioWorkspace.cs]: There's a comment in the else about one case, which is AdditionalDocuments. @mavasani mentioned something in the bug about being resilient against documents that had been removed from the solution, but I don't know exactly how that happens.

File: src/VisualStudio/Core/Impl/RoslynVisualStudioWorkspace.cs [hunk 2]
```diff
@@ -124,8 +124,22 @@ internal override IInvisibleEditor OpenInvisibleEditor(IVisualStudioHostDocument
...
+                {
+                    // Disable undo on generated documents
+                    needsUndoDisabled = this.Services.GetService<IGeneratedCodeRecognitionService>().IsGeneratedCode(this.CurrentSolution.GetDocument(hostDocument.Id));
+                }
+                else
+                {
+                    // Enable undo on "additional documents" or if no document can be found.
+                    needsUndoDisabled = false;
```
Comment (balajikris) [inline src/VisualStudio/Core/Impl/RoslynVisualStudioWorkspace.cs]: This is already false, are you setting again for clarity of the code flow?

**Comment by balajikris** [comment]
:+1:  for the change. May be invoking "test vsi please" might offer some generic testing for this.

File: src/VisualStudio/Core/Impl/RoslynVisualStudioWorkspace.cs (hunk 2 above)
Comment (dpoeschl) [inline src/VisualStudio/Core/Impl/RoslynVisualStudioWorkspace.cs]: Yes. I could remove the ```else``` and keep the clarity with a different comment, but this felt like the most readable approach.

**Comment by dpoeschl** [comment]
test this please

**Pull request merged**